Minor test framework fixes

delcypher opened the pull request:
These fixes attempt to unify some of the differences between the way tests are built between the CMake and Makefile build systems and is a necessary step towards doing #858.

In doing this work I observed that building of the renderscript tests is broken which @aam has addressed via #890 . This support isn't in the CMake build yet so once this and #890 are merged I'll implement that.

abadams commented:
For the record, I still disagree with #858 - I'm really not into adding another scripting language as a dependency.

But unifying the binary names between the two build systems is reasonable. I think test/scripts/test_target.sh needs a fix (it has a make test_argmax in it).

abadams commented:
Also, I'd like to delay this and other build system refactorings until after the siggraph course (next week). I want a reliable master branch and a solid release for newcomers, and build system churn introduces minor bugs that break the nightlies in obscure corner cases.

delcypher commented:
> For the record, I still disagree with #858 - I'm really not into adding another scripting language as a dependency.

Ok. At the moment I'm planning to make (python) optional. There's nothing stopping someone running the tests by hand but running with ``lit`` will be much faster and more convenient.

I still don't think python is a particularly onerous requirement for running the tests. All platforms I'm aware of will have it installed.

- Linux/OSX have it installed by default
- Windows. The LLVM binary releases for Windows can't be used by Halide (I tried). Various tools were missing in the binary release (3.6) so a user must compile LLVM from source which means they **must** have python because it is a hard dependency for building LLVM.

> But unifying the binary names between the two build systems is reasonable. I think test/scripts/test_target.sh needs a fix (it has a make test_argmax in it).

I've incorporated a fix for this.

delcypher commented:
> Also, I'd like to delay this and other build system refactorings until after the siggraph course (next week). I want a reliable master branch and a solid release for newcomers, and build system churn introduces minor bugs that break the nightlies in obscure corner cases.

Isn't that what proper releases are for? The master branch of most projects is inherently unstable. Rather than slowing down the merging of various features wouldn't it be better to make a tag or branch for the version you want used for your siggraph course? This would also mean you have a human readable name you can refer to in your slides/course notes which will be relevant for a very long time (provided we don't delete the branch/tag) whereas the state of ``master`` is always in flux.

You can easily tell people to use a different branch when cloning.
```
$ git clone -b your_siggraph2015_branch https://github.com/halide/Halide.git
```

Of course this is up to you. I can wait for these changes to be merged.

abadams commented:
I am indeed going to tell people to download the release I'm about to cut
or the corresponding tag, however I expect some people will just jump in
with master, so I'd like master to be more stable than usual over the next
two weeks.

I'm sure you've noticed I'm generally averse to change when it comes to the
build system. This is because every time there's a change I spend an hour
or two every morning over the next week fixing corner cases that failed
overnight, and this is not work I enjoy doing, and not work I can do while
at siggraph. Those corner case failures often mask more serious problems,
like new bugs.

So build system changes really need to substantially reduce future
maintenance, or add a very compelling feature. Given that the code in the
Makefile and powershell script that runs the tests works and is about as
simple as it could be, I don't buy the need for an additional python
script, unless it has features I'm not anticipating. It seems like it will
be a net addition of code that duplicates existing functionality and adds a
new programming language to the project.



On Tue, Aug 4, 2015 at 11:46 PM, Dan Liew <notifications@github.com> wrote:

> Also, I'd like to delay this and other build system refactorings until
> after the siggraph course (next week). I want a reliable master branch and
> a solid release for newcomers, and build system churn introduces minor bugs
> that break the nightlies in obscure corner cases.
>
> Isn't that what proper releases are for? The master branch of most
> projects is inherently unstable. Rather than slowing down the merging of
> various features wouldn't it be better to make a tag or branch for the
> version you want used for your siggraph course? This would also mean you
> have a human readable name you can refer to in your slides/course notes
> which will be relevant for a very long time (provided we don't delete the
> branch/tag) whereas the state of master is always in flux.
>
> You can easily tell people to use a different branch when cloning.
>
> $ git clone -b your_siggraph2015_branch https://github.com/halide/Halide.git
>
> Of course this is up to you. I can wait for these changes to be merged.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/892#issuecomment-127891718>.
>

delcypher commented:
@abadams I've rebased on master, fixed merged conflicts and also now that the renderscript tests build I have

- Added support for building renderscript tests to CMake
- Added the building of renderscript tests to the ``build_tests`` target for the Makefile build system

abadams commented on test/CMakeLists.txt:
```diff
@@ -47,7 +64,7 @@ if (WITH_TEST_GENERATORS)
   file(GLOB TESTS RELATIVE "${CMAKE_CURRENT_SOURCE_DIR}/generator" "${CMAKE_CURRENT_SOURCE_DIR}/generator/*_jittest.cpp")
   foreach(test ${TESTS})
-    string(REPLACE "_jittest.cpp" "" name "generator_jit_${test}")
+    string(REPLACE "_jittest.cpp" "" name "generator_exec_jit_test_${test}")
```
This is a dumb problem, but...

The windows buildbots currently have problems with really long file names (>260 characters). Indeed, I'm currently running a build slave in a directory called E:/b/ to avoid this. The problem seems to be that generators end up in deeply nested folders that contain the name of the generator twice. Perhaps this change could be modified to not extend the length of any filenames?

steven-johnson replied:
This is a dumb response, but...

Surely Windows systems in 2015 can be configured to avoid this Neanderthal limitation?

abadams replied:
Perhaps, but a default windows 8 install has this problem, and I don't want
to tell users to change their file system configuration to build Halide :)

On Thu, Aug 13, 2015 at 11:55 AM, Steven Johnson <notifications@github.com>
wrote:

> In test/CMakeLists.txt
> <https://github.com/halide/Halide/pull/892#discussion_r37011619>:
>
> > @@ -47,7 +64,7 @@ if (WITH_TEST_GENERATORS)
> >    # and make an executable target for it named 'generator_foo'
> >    file(GLOB TESTS RELATIVE "${CMAKE_CURRENT_SOURCE_DIR}/generator" "${CMAKE_CURRENT_SOURCE_DIR}/generator/*_jittest.cpp")
> >    foreach(test ${TESTS})
> > -    string(REPLACE "_jittest.cpp" "" name "generator_jit_${test}")
> > +    string(REPLACE "_jittest.cpp" "" name "generator_exec_jit_test_${test}")
>
> This is a dumb response, but...
>
> Surely Windows systems in 2015 can be configured to avoid this Neanderthal
> limitation?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/892/files#r37011619>.
>

kgnk replied:
You would think that 256 byte (not character) path length (not including drive name and terminating NUL) would be a stupid limitation, but it is a carry over from the BIOS call that returns the length in an 8-bit register. Microsoft doesn't like to deprecate anything, ever, if it can possibly be avoided, so applications written 20 years ago can still work. Fixing it means touching code all over the build tools that assume a buffer of 260 bytes is enough. On the other hand, Apple have changed CPUs twice (Moto 68K -> Power -> Intel x86), totally rebased the OS and apps that were written for the previous OS update may stop working.

delcypher commented:
> The windows buildbots currently have problems with really long file names (>260 characters). Indeed, I'm currently running a build slave in a directory called E:/b/ to avoid this. The problem seems to be that generators end up in deeply nested folders that contain the name of the generator twice. Perhaps this change could be modified to not extend the length of any filenames?

**WINDOWS! \*\*Shakes fist at sky\*\*!** So lame...

I'll take a look and try to come up with a naming scheme that has short names.

delcypher commented:
@abadams I've rebased on master and I've made the test executable names a little shorter. I've also changed the name of the generator executables `` generator_foo -> foo.generator`` to match the Makefile build system and for clarity.

abadams commented:
I have a experimental buildbot setup going based on Alex's stuff: http://104.154.46.123:8010/waterfall

I'm going to get those builds green (just one minor issue on arm left), and then use this PR to test testing PRs on all platforms using buildbot.

aam commented:
That is quite serious buildbot farm!

On Tue, Aug 18, 2015 at 11:03 AM Andrew Adams <notifications@github.com>
wrote:

> I have a experimental buildbot setup going based on Alex's stuff:
> http://104.154.46.123:8010/waterfall
>
> I'm going to get those builds green (just one minor issue on arm left),
> and then use this PR to test testing PRs on all platforms using buildbot.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/892#issuecomment-132299683>.
>

abadams commented:
I'm planning on replacing the nightly builds with it - it uploads releases
here: http://104.154.46.123/

On Tue, Aug 18, 2015 at 12:19 PM, Alexander Aprelev <
notifications@github.com> wrote:

> That is quite serious buildbot farm!
>
> On Tue, Aug 18, 2015 at 11:03 AM Andrew Adams <notifications@github.com>
> wrote:
>
> > I have a experimental buildbot setup going based on Alex's stuff:
> > http://104.154.46.123:8010/waterfall
> >
> > I'm going to get those builds green (just one minor issue on arm left),
> > and then use this PR to test testing PRs on all platforms using buildbot.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/halide/Halide/pull/892#issuecomment-132299683>.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/892#issuecomment-132323046>.
>

aam commented:
Cool, sounds good!

abadams commented:
I pulled this PR into a branch, and then triggered a forced build of it on Windows here: 

http://104.154.46.123:8010/builders/win-trunk/builds/110

abadams commented:
Looks like there are some windows errors to fix now that renderscript is on. I'll fix them in the branch and restart the windows test.

abadams commented:
Restarted with fixes: http://104.154.46.123:8010/builders/win-trunk/builds/111
